Overhaul Rules

oxinabox opened the pull request:
This is a very mighty PR.
 
 - Adds a extra (currently always`DNE()` as not functors) return value for all `rrules`, to represent the derviative w.r.t internals of closures/functors, and similar demands an extra input argument at the start of a call to `frule`  (ignored for all current cases as not functors)  #22
 - the propergators (pushback/pullforward) now always return a tuple, even if it has just 1 element. #31
 -  `frule`/`rrule`  now return a 1 propagator (pushforward/pullback) that returms a tuple of partials, rather than 1 propagator per partial  #38 
 - as a result `AbstractRule` subtypes are no longer used anywhere  #39  
 - `@scalar_rule` automatically names pullbacks/pushforwards. Below is what that looks like in Julia Master (with new improved display for gensymed names)

```
julia> _, pushforward = frule(sin, 1)
(0.8414709848078965, ChainRules.var"##75#sin_pushforward#55"{Int64}(1))
```

Does not look quiet as nice for 1.0 but still useful
```
julia> _, pushforward = frule(sin, 1)
(0.8414709848078965, getfield(ChainRules, Symbol("##75#sin_pushforward#55")){Int64}(1))
```

It has a corresponding PR to ChainRules.jl

https://github.com/JuliaDiff/ChainRules.jl/pull/91


This is the main blocker for  https://github.com/FluxML/Zygote.jl/pull/291

oxinabox commented:
All the accumulation stuff needs to be rewritten still.

oxinabox commented:
I have not delted the AbstractRules yet, as I am yet to workout the story for `store!`
and for 2 arg rules that know how to `update!`.

simeonschaub commented:
@oxinabox  I've created a test case for mixed Wirtinger derivatives [here](https://gist.github.com/simeonschaub/a6dfcd71336d863b3777093b3b8d9c97). This should help making sure, that this works correctly.

oxinabox commented:
Cool. I will add that to this PR tomorrow.
At least to check all the returned types are right.

simeonschaub commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,73 @@
+    end
+end
+
+function standard_propergation_expr(Δs, ∂s)
```
```suggestion
function standard_propagation_expr(Δs, ∂s)
```

simeonschaub commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,73 @@
+    return :(+($(∂_mul_Δs...)))
+end
+
+function wirtinger_propergation_expr(𝒟, wirtinger_indices, Δs, ∂s)
```
```suggestion
function wirtinger_propagation_expr(𝒟, wirtinger_indices, Δs, ∂s)
```

simeonschaub commented on src/rule_definition_tools.jl:
```diff
@@ -34,11 +107,16 @@ Constraints may also be explicitly be provided to override the `Number` constrai
+At present this does not support defining for closures/functors.
+Thus in reverse-mode, the first returned partial,
+representing the derivative with respect to the function itself, is always `NO_FIELDS`.
+And in forwards-mode, the first input to the returned propagator is always ignored.
```
```suggestion
And in forward-mode, the first input to the returned propagator is always ignored.
```

simeonschaub commented on src/differentials.jl:
```diff
@@ -206,3 +206,27 @@ end
+    differential(𝒟::Type, der)
+
+For some differential (e.g. a `Number`, `AbstractDifferential`, `Matrix` etc),
+converts it to another differential that is more suited for the domain given by
```
```suggestion
convert it to another differential that is more suited for the domain given by
```
 Imperative form sounds better to me here.

simeonschaub commented on src/differentials.jl:
```diff
@@ -206,3 +206,27 @@ end
+"""
+    differential(𝒟::Type, der)
+
+For some differential (e.g. a `Number`, `AbstractDifferential`, `Matrix` etc),
```
```suggestion
For some differential (e.g. a `Number`, `AbstractDifferential`, `Matrix`, etc.),
```

simeonschaub commented on src/differentials.jl:
```diff
@@ -206,3 +206,27 @@ end
+"""
+    NO_FIELDS
+
+Constant for the reverse-mode derivative with respect to structure that has no fields.
```
```suggestion
Constant for the reverse-mode derivative with respect to a structure that has no fields.
```

simeonschaub commented on test/runtests.jl:
```diff
@@ -2,7 +2,7 @@
 using ChainRulesCore
 using LinearAlgebra: Diagonal
-using ChainRulesCore: extern, accumulate, accumulate!, store!, @scalar_rule,
+using ChainRulesCore: extern, accumulate, accumulate!, store!,  @scalar_rule,
```
```suggestion
using ChainRulesCore: extern, accumulate, accumulate!, store!, @scalar_rule,
```

oxinabox commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,73 @@
+    if it is taken at `1+1im` it returns `Complex{Int}`.
+    At present it is ignored for non-Wirtinger derivatives.
+"""
+function propergation_expr(𝒟, Δs, ∂s)
```
```suggestion
function propagation(𝒟, Δs, ∂s)
```

oxinabox commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,73 @@
+    if isempty(wirtinger_indices)
+        return standard_propergation_expr(Δs, ∂s)
+    else
+        return wirtinger_propergation_expr(𝒟, wirtinger_indices, Δs, ∂s)
```
```suggestion
        return wirtinger_propagation_expr(𝒟, wirtinger_indices, Δs, ∂s)
```

oxinabox commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,73 @@
+    end
+    ∂s = map(esc, ∂s)
+    if isempty(wirtinger_indices)
+        return standard_propergation_expr(Δs, ∂s)
```
```suggestion
        return standard_propagation_expr(Δs, ∂s)
```

oxinabox commented on src/differentials.jl:
```diff
@@ -206,3 +206,27 @@ end
+
+For some differential (e.g. a `Number`, `AbstractDifferential`, `Matrix`, etc.),
+convert it to another differential that is more suited for the domain given by
+the type 𝒟.
```
Prob should say "convert if required" or some such.

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -59,6 +137,9 @@ For examples, see ChainRulesCore' `rules` directory.
 macro scalar_rule(call, maybe_setup, partials...)
+    ############################################################################
+    # Setup: normalizing input form etc
+
```
Can this be broken up into functions? I'd love for this to not be 100 lines long...

nickrobinson251 replied:
i still feel the same

But if it is not easy / does not make sense to you, @oxinabox , then that's fine by me too

nickrobinson251 replied:
Wahooo! Thanks 🎉

nickrobinson251 commented on test/rules.jl:
```diff
@@ -31,8 +31,42 @@ _second(t) = Base.tuple_type_head(Base.tuple_type_tail(t))
-    @test fr(1) == 1
-    rrx, rr = rrule(cool, 1)
+    @test fr(NamedTuple(), 1) == (1,)
+    rrx, (rr) = rrule(cool, 1)
```
why the brackets around `rr`?

oxinabox replied:
Left over from when things were different

```suggestion
    rrx, rr = rrule(cool, 1)
```

nickrobinson251 commented on test/rules.jl:
```diff
@@ -31,8 +31,42 @@ _second(t) = Base.tuple_type_head(Base.tuple_type_tail(t))
+    f, pushforward = frule(myabs2, x)
+    @test f === x^2
+
+    df = @inferred pushforward(NamedTuple(), One())
```
minor, but I was confused by this test, 
becuase at first I thought `pushforward` was a generic `ChainRulesCore` function and couldn't see how it was getting info about `f = myabs2`... can we rename it `f_pushforward` or `myabs2_pushforward` or somesuch?

I know it's only test code, but I often read tests for understanding how the package is meant to be used :)

nickrobinson251 replied:
by the way, naming gripe aside, this testset is a great little tutorial!

nickrobinson251 commented on src/rule_types.jl:
```diff
@@ -51,12 +51,15 @@ See also: [`frule`](@ref), [`rrule`](@ref), [`Rule`](@ref), [`DNERule`](@ref), [
 # in order to get the `i`th rule (assuming it's 1)
-Base.getindex(rule::AbstractRule, i::Integer) = i == 1 ? rule : throw(BoundsError())
+function Base.getindex(rule::AbstractRule, i::Integer)
+    @warn "iterating rules is going away"
```
We still need to figure out hwo to deprecate `AbstractRule`s altogether? I guess you'll figure that out when/once `acummulate!` etc get rewritten?

oxinabox replied:
I'm not too worried about deprecating them
as the changes are completely breaking.
And it isn't used in too many places.

But do need to workout how to remove them outright

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
+        if Meta.isexpr(partial, :tuple)
+            partial
+        else
+            @assert length(inputs) == 1
```
is there anyway this could ever fail? I mean, i presume not, otherwise you wouldn't make it an assert... but just double checking!

oxinabox replied:
It actually could fail, 
I should change it it was an assertion before this PR and I just moved it,
but I might was well fix it on the way past.

```suggestion
            length(inputs) == 1 || error("Invalid use of `@scalar_rule`")
```

I'm OK with using `error` here as it can't be recovered from and is a programemer mistake

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
+    # Main body: defining the results of the frule/rrule
+
+    # An expression that when evaluated will return the type of the input domain.
+    # Multiple repetitions of this expression should optimize ot. But if it does not then
```
typo `ot` -> `out`
```suggestion
    # Multiple repetitions of this expression should optimize out. But if it does not then
```

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
+    𝒟 = :(typeof(first(promote($(call.args[2:end]...)))))
+
+    n_outputs = length(partials)
+    n_inputs = length(inputs)
```
Since we are computing `n_inputs` here anyway, can we compute it earlier and use it above when making `partials` a tuple? Can't see exactly how we'd re-write it -- but just a thought.

oxinabox replied:
Do you mean for theck of `length(inputs)==1` ?
yeah maybe

idea of `n_inputs` (and `n_outputs`) is to be clear on shape of what is going on with in the pullback and pushforward definitions and what lines up with what.

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
-    return quote
+
+    ############################################################################
+    # Final return: building the expression to insert in the place of this macro
```
i'd still prefer this were more modular... but I really appreciate how well the code reads - good work!

nickrobinson251 commented on src/differentials.jl:
```diff
@@ -206,3 +206,27 @@ end
+####
+"""
+    differential(𝒟::Type, der)
+
```
I'm a little confused by what role this plays in the API

I see it's use in `[wirtinger_]propagation_expr`, but when would it be used otherwise?

oxinabox replied:
At present it is only used in 
`wirtinger_propagation_expr` it does the things 
where if you know you are in a real domain (because the input point was real)
and you know the wirtinger derviative, then convert that to a real derviative.

I can imagine us doing other things with it later.
Some speculations:
 - We might accept other forms of generalized derivatives (What is Wirtinger for Quaternions?)
 - We might like to do `differential(::Type{<:Number}, der::Number) = iszero(der) ? Zero() : der`
 
But that is out of scope of this PR.

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,73 @@
 # These are some macros (and supporting functions) to make it easier to define rules.
 
+propagator_name(f::Expr, propname::Symbol) = propagator_name(f.args[end], propname)
```
can we attach a docstring / comment / example to these please?

oxinabox replied:
```suggestion
"""
    propagator_name(f, propname)
Determines a reasonable name for the propagator function.
The name doesn't really matter too much as it is a local function to be returned
by `frule` or `rrule`, but a good name make debugging easier.
`f` should be some form of AST representation of the actual function,
`propname` should be either `:pullback` or `:pushforward`

This is able to deal with fairly complex expressions for `f`:

    julia> propagator_name(:bar, :pushforward)
    :bar_pushforward

    julia> propagator_name(esc(:(Base.Random.foo)), :pullback)
    :foo_pullback

propagator_name(f::Expr, propname::Symbol) = propagator_name(f.args[end], propname)
```

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
+        # 1 partial derivative per input
+        pullback_returns = map(1:n_inputs) do input_i
+            ∂s = [partial.args[input_i] for partial in partials]
+            propergation_expr(𝒟, Δs, ∂s)
```
```suggestion
            propagation_expr(𝒟, Δs, ∂s)
```

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
+        Δs = [Symbol(string(:Δ, i)) for i in 1:n_inputs]
+        pushforward_returns = map(1:n_outputs) do output_i
+            ∂s = partials[output_i].args
+            propergation_expr(𝒟, Δs, ∂s)
```
```suggestion
            propagation_expr(𝒟, Δs, ∂s)
```

simeonschaub commented on src/differentials.jl:
```diff
@@ -206,3 +206,27 @@ end
+
+####
+"""
+    differential(𝒟::Type, der)
```
Maybe this should take `primal` and `conjugate` as arguments, and depending on `𝒟` return either `Wirtinger` or their sum? I think that would make it more clear to rule authors, that when you create a `Wirtinger`, you usually also want this fall-through behavior.

simeonschaub replied:
It could also be useful to check, whether `conjugate isa Zero` here, and unwrap `Wirtinger` if that's the case.

oxinabox replied:
>Maybe this should take primal and conjugate as arguments, and depending on 𝒟 return either Wirtinger or their sum?

I'm not sure I understand the advantag of that?

With this we have
`differential(𝒟, Wirtinger(primal, conjugatge))`
which seems fine.
What is the advantage of
`differential(𝒟, primal, conjugatge)` 
?

> It could also be useful to check, whether conjugate isa Zero here, and unwrap Wirtinger if that's the case.

Maybe.
Maybe even `iszero(conjugate)` to to get constants like `0`

Maybe even for inputs that are scalar:
`iszero(der) && return Zero()`
and similar for `One()`

We should discuss that kind thing in an issue and make a follow up PR for it

simeonschaub replied:
> What is the advantage of
> `differential(𝒟, primal, conjugatge)`
> ?

I just don't know if `differential` is the best name for this function, since it takes a differential and returns a, sometimes different and more efficient, differential again. I would expect a function called `differential` to work more like a constructor. Or do we maybe want to call this `wirtinger` and have it take `𝒟`, `primal`, and `conjugate`, since this probably corresponds better to what it does right now? But I also wouldn't feel too strongly just leaving this for now, since I'm also struggling to find a better name for this function.

> Maybe.
> Maybe even `iszero(conjugate)` to to get constants like `0`
> 
> Maybe even for inputs that are scalar:
> `iszero(der) && return Zero()`
> and similar for `One()`

I'm not quite convinced we benefit from introducing dispatch based on value here, wouldn't this also cause problem on GPUs? But this is definitely an issue for another day.

oxinabox replied:
How about we rename it to `refine_differential` ?
I think this actually also interacts with #8  since we will want to apply something recursively.

simeonschaub commented on src/differentials.jl:
```diff
@@ -206,3 +206,27 @@ end
+convert it to another differential that is more suited for the domain given by
+the type 𝒟.
+"""
+function differential(::Type{<:Union{<:Real, AbstractArray{<:Real}}}, w::Wirtinger)
```
Maybe check for `eltype` instead, to allow for iterables that are not `<: AbstractArray`?

oxinabox replied:
right now the idea of having other iterators as differentials is not something I've fully worked though in my head.

simeonschaub replied:
Ok, this should be fine for now. There's definitely still a lot to do for supporting `Wirtinger` of arrays correctly.

simeonschaub commented on test/rules.jl:
```diff
@@ -31,8 +31,42 @@ _second(t) = Base.tuple_type_head(Base.tuple_type_tail(t))
     frx, fr = frule(cool, 1)
     @test frx == 2
-    @test fr(1) == 1
+    @test fr(NamedTuple(), 1) == (1,)
```
If I understand correctly, the first argument is just ignored. Should we just make it convention to pass an empty tuple here instead, if we don't need this? `NamedTuple()` seems a bit long and potentially confusing to me.

oxinabox replied:
Fair point, for context:

If ever it is not ignored, because it has been overloaded,
because the thing is a functor,
then it will in the real world always be a `NamedTuple`, but with some fields.

I can see the arguement to use `()` for the case we know it will be ignored.
Someone else want to chime in,

We probably do not want to use this "shorthand".
```
julia> (;()...)
NamedTuple()
```

omus commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,89 @@
 # These are some macros (and supporting functions) to make it easier to define rules.
 
+"""
+    propagator_name(f, propname)
```
```suggestion
    propagator_name(f, propname)

```

omus commented on src/differentials.jl:
```diff
@@ -206,3 +206,28 @@ end
+"""
+const NO_FIELDS = DNE()
+
+####
```
```suggestion
```

omus commented on src/differentials.jl:
```diff
@@ -206,3 +206,28 @@ end
 Base.conj(x::Thunk) = @thunk(conj(extern(x)))
+
+
+####
```
```suggestion
```

omus commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,89 @@
+    For example if the derivative is being taken at the point `1` it returns `Int`.
+    if it is taken at `1+1im` it returns `Complex{Int}`.
+    At present it is ignored for non-Wirtinger derivatives.
+"""
```
Content is overly indented?

oxinabox replied:
Yes, good catch

tkf commented:
> * `frule`/`rrule`  now return a 1 propagator (pushforward/pullback) that returms a tuple of partials, rather than 1 propagator per partial  #38

I have a concern about this decision.  Does it mean that, for example, the reverse rule of `*` becomes something like (computationally equivalent to)

```julia
function rrule(::typeof(*), A::AbstractMatrix{<:Real}, B::AbstractMatrix{<:Real})
    return A * B, Rule(Ȳ -> (Ȳ * B', A' * Ȳ))
end
```

(which is like [what Zygote.jl is doing at the moment](https://github.com/FluxML/Zygote.jl/blob/d74f3cf5ed3c185969ff7787ee36d15c105e6653/src/lib/array.jl#L207-L209)) instead of [the current definition](https://github.com/JuliaDiff/ChainRules.jl/blob/d8339cc1ecd8b8180dfb4f6797b168d287b5e92c/src/rulesets/LinearAlgebra/dense.jl#L75-L77)

```julia
function rrule(::typeof(*), A::AbstractMatrix{<:Real}, B::AbstractMatrix{<:Real})
    return A * B, (Rule(Ȳ -> Ȳ * B'), Rule(Ȳ -> A' * Ȳ))
end
```

where you can compute the derivative w.r.t different argument separately?

Wouldn't it be a huge performance loss when large constant arrays are participated in the computation of the intermediate variables that depend on the variables ("trainable parameters") with which the derivatives are taken?  For example, in the Generative Adversarial Network (GAN) setting, I think it would be a big issue when taking derivative `(d/dp) D(G(p))` w.r.t the parameter `p` of the generator `G` while treating the parameters of discriminator `D` as constant.  I also noted the concerns in other similar situations here https://github.com/FluxML/Zygote.jl/issues/323#issuecomment-527627148.

I'm by no means an AD or ML specialist so I may be missing something.  It would be great if you can clarify that my concern is invalid.

oxinabox commented:
**Short answer:** don't worry we solve this with `Thunks`.

---

#### Full answer:

@tkf a very reasonable concern. And one I used too have 
until I understood why whe have the `Thunk` differential.
(https://github.com/JuliaDiff/ChainRulesCore.jl/issues/18)
We have an issue onpen about documenting that a bit more.
https://github.com/JuliaDiff/ChainRulesCore.jl/issues/46





> ... [the current definition](https://github.com/JuliaDiff/ChainRules.jl/blob/d8339cc1ecd8b8180dfb4f6797b168d287b5e92c/src/rulesets/LinearAlgebra/dense.jl#L75-L77)
> 
> ```julia
> function rrule(::typeof(*), A::AbstractMatrix{<:Real}, B::AbstractMatrix{<:Real})
>     return A * B, (Rule(Ȳ -> Ȳ * B'), Rule(Ȳ -> A' * Ȳ))
> end
> ```


becomes: [In partner PR, this is one of the ones I've already updated](https://github.com/JuliaDiff/ChainRules.jl/pull/91/files#diff-e59f5e7b34490be11e2a82fdb8b831edR86)

```
function rrule(::typeof(*), A::AbstractMatrix{<:Real}, B::AbstractMatrix{<:Real})
    return A * B, Ȳ -> (NO_FIELDS, @thunk(Ȳ * B'), @thunk(A' * Ȳ))
end
```

` @thunk(Ȳ * B')` is just a shorthand for `Thunk(()->Ȳ * B')`
`Thunk` source: [definition](https://github.com/JuliaDiff/ChainRulesCore.jl/blob/355b8306d05ff2efbc80964e4f7f4a5ab04dc403/src/differentials.jl#L180-L208), and [math](https://github.com/JuliaDiff/ChainRulesCore.jl/blob/355b8306d05ff2efbc80964e4f7f4a5ab04dc403/src/differential_arithmetic.jl#L92-L100)

That it basically is is a differential that defers computation until it is used.
So if it is never used then the wrapped computation is never computed
For examole:
```
Y, pullback = rrule(*, A, B)
_, dA_diff, dB_diff = pullback(One()
dB = extern(dB_diff)
```
then since `dA_diff` was never `extern`ed the `Ȳ * B'` is never evaluated.

tkf commented:
@oxinabox Thanks a lot! I appreciate the full explanation. I should have checked the partner PR.

oxinabox commented:
> @oxinabox Thanks a lot! I appreciate the full explanation. I should have checked the partner PR.

It is a really important question

mattBrzezinski commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,90 @@
+            ∂f̄∂ī_mul_Δ̄ = :(conj(wirtinger_primal($∂i)) * wirtinger_conjugate($Δi))
+            push!(∂_mul_Δs_primal, :($∂f∂i_mul_Δ + $∂f∂ī_mul_Δ̄))
+            push!(∂_mul_Δs_conjugate, :($∂f̄∂i_mul_Δ + $∂f̄∂ī_mul_Δ̄))
+        else
```
You append to `∂_mul_Δs_primal` and `∂_mul_Δs_conjugate`, but nothing gets entered into `∂_wirtinger_defs`.

Below in the `return` statement we splat `∂_wirtinger_defs`, how will the above (missing values in `∂_wirtinger_defs`) be handled in this return statement? Will the corresponding value for `∂_wirtinger_defs` be `null`? Does it matter?

oxinabox replied:
I am not sure, this logic has not been touched in this PR,
and I have never read it closely.
I just moved it around into a new function

oxinabox replied:
I think to will not get added to output at all.

I think it can just be deleted in any case

oxinabox replied:
Actually I see it is used in the other side of the `if`, so this is needed and it is fine to have it as an empty array, it will just not turn up in the returned expression

jrevels commented:
> frule/rrule now return a 1 propagator (pushforward/pullback) that returms a tuple of partials, rather than 1 propagator per partial

I missed the discussion on this. Why make this change?

There are two points here:

- Right now (or, well, when I originally designed the package) `frule`/`rrule` were designed to be duals of each other to facilitate composing mixed mode code. It's also internally consistent: `*rule` returns the output value and a rule for each requested differential (the "requested differentials" being output perturbations in forward-mode, and input sensitivities in reverse-mode). 

- I don't understand the work sharing point - either approach allows you to close over shared work when desired, correct? However, the original design has an extra advantage: it makes it easier/more natural to elide unnecessary work (i.e. if my reverse propagation doesn't flow through a specific node, that work need not be computed; similar to [my explanation of Thunks](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/18#issuecomment-519211209) but it's a different stage). This is true regardless of mode, really - it's a nice property of returning "a rule for each requested differential".

It seems like this change would sacrifice both of the above advantages. Is that not the case, or is there a significant overriding benefit that's worth it?

jrevels commented:
Just saw https://github.com/JuliaDiff/ChainRulesCore.jl/pull/30#issuecomment-527821693; thanks for the explanation. I see now that you're planning on staging that work separately via `@thunk`, but I'm still unclear on what the motivation is behind the change. Why is manually `@thunk`ing the right thing to do here vs. encoding the separation of concerns/work via the rule structure?

oxinabox commented:
I have written a short book, explaining, 
which is fine some of this I will reformulate into docs



Consider some rrule, for a function with two inputs.
where the partials share some common computation
represented by `buzz(a, b)`

### With this PR:
```
function rrule(::typeof(foo), a, b)
    y = foo(a, b)
    function foo_pullback(ȳ)
        h = buzz(a, b)

        ∂a = @thunk(2h)
        ∂b = @thunk(2b)
        return NO_FIELDS, ∂a, ∂b
    end
    return y, y_pullback
end

_, pullback = rrule(foo, 1, 2)
_, da, db = pullback(1)  # buzz was only called once. on this line
```

### Without this PR, Option one:
Pay-for it even when only wanting the main result:
```
function rrule(::typeof(foo), a, b)
    y = foo(a, b)
    h = buzz(a, b)
    ∂a = Rule(ȳ -> 2h)
    ∂b = Rule(ȳ -> 3h)
    return ∂a, ∂b
    return y, (∂a, ∂b)
end

_, da_pb, db_pb = rrule(foo, 1, 2)  # buzz was only called once. on this line
da = da_pb(1)
db = db_pb(1)
```

### Without this PR, Option two:
Pay-for it per each partial.
(This  can also be written by thunking `h`)
```
function rrule(::typeof(foo), a, b)
    y = foo(a, b)
    ∂a = Rule(ȳ -> 2buzz(a, b))
    ∂b = Rule(ȳ -> 3buzz(a, b))
    return ∂a, ∂b
    return y, (∂a, ∂b)
end


_, da_pb, db_pb = rrule(foo, 1, 2)
da = da_pb(1)  # buzz was called on this line
db = db_pb(1)  # buzz was called again on this line
```
---


Something like this actually occurs for some of the `/` rules.``
https://github.com/JuliaDiff/ChainRules.jl/pull/91/files#diff-e59f5e7b34490be11e2a82fdb8b831edR105-R123
though that case is even more complex as somethings are shared between
the partials, and some between both the partials and the main rusult.
(but for the stuff related to partials it resolve it by doing the share work (the computation of `dC` twice))

This doesn't solve the issue entirely,
since for things with say 3 partials where `buzz` is only shared between two of them.
(as is infact the case for the `NO_FIELDS` partial)

If one only ever wants 1 partial, then this has no advantage over many rules,
and is probably worse.
but if one wants multiple then this sharing becomes useful.

---

There are several other things also:

As we start to handle derivatives of structured return types.
Like `svd` returning a `SVD` struct, or even something that takes in (rrule) or returns (frule) a tupple of tuples.
this approach is consistent when applied recursively.
1 pullback returns a tuple of differential, which may be made up of differentials which may be ... and so forth.
where-as without this it is:
a tuple of pullbacks (Rules) each of which returns is a differential, which may be made up of differentials which may be ... 
so the outermost layer is different.
Which is not too bad really. I think for `frule` of a function returing a tuple of tuples it is where is the worse.
(since just does recursively destructure nested tuples during assignment)

Getting rid of the Rules concept *really* simplifies this for people contributing.
Without this PR you have 3 new types/important functions:
`rrule`/`frule`, `AbstractDifferential`, `AbstractRule`.
The last is the one that is the hardest. I was certainly confuse for a while.
At first it looks a lot like `AbstractDifferential`,
since there are `AbstractRules` for many of the AbstractDifferentials like `DNERule`/`DNE`, `ZeroRule`/`Zero`
and `Rule`/`Thunk`.
You have to learn that `AbstractRules` are special kinds of functions (but special in a different way to `Thunk`s).
and that you need one per input for `rrules` and one per output for `ffules`.
and they don't actually do much that is special (since we got rid of Cassette).
Now both `rrule` and `frule` return exactly one of this pullback/pushward function.
and the pullback takes some difference (wobble) in output space, and give back some differences (wiggle) in input space.
where as the pushforward takes some difference in input space (wiggle) and gives back some difference in output space (wabble).
So those things each just map to ther core of what `rrule` and `frule` do.
Whereas I find describing the collections of `Rules` that do the same, harder.

---

I am not sure I really understand  what you mean when you say
that `frule` and `rrule` are duels, to help support mixed-mode AD.
I don't really know that stuff, to be completely honest.
I don't think we would have broken any elegence because we turned both
`rrule`s and `frule`s use of rules inside out.
So they should still match-up.


The following might be of interest:

```
julia> _, pushforward = frule(sin, 1)
(0.8414709848078965, getfield(ChainRules, Symbol("##75#sin_pushforward#55")){Int64
}(1))

julia> _, pullback = rrule(sin, 1)
(0.8414709848078965, getfield(ChainRules, Symbol("##79#sin_pullback#57")){Int64}(1
))

julia> pullback(0.8)
(DNE(), 0.4322418446945118)

julia> pushforward(pullback(0.8)...)
0.2335412653811431

julia> pullback(pushforward(pullback(0.8)...)...)
(DNE(), 0.12618288420079476)
```

oxinabox commented on src/differentials.jl:
```diff
@@ -199,6 +201,22 @@ Thunk(var"##8#10"())
+By `@thunk`ing the work required for each, they can only be computed when needed.
+
+#### So why not thunk everything?
+`@thunk` creates a closure over the expression, which is basically a struct with a
```
i like it.

oxinabox commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,90 @@
+    # Notice: the thunking of `∂s[i] (potentially) saves us some computation
+    # if `Δs[i]` is a `AbstractDifferential` otherwise it is computed as soon
+    # as the pullback is evaluated
+    ∂_mul_Δs = [:(@thunk($(∂s[i])) * $(Δs[i])) for i in 1:length(∂s)]
```
You should get an error.
Right now it will be a bounds error.

oxinabox replied:
Can we leave this as is, since it isn't changed in this PR?

nickrobinson251 replied:
Yes :) Is it worth opening an issue (to stare hard at this and figure out if all is well)?

jrevels commented:
Thanks, that's a very nice explanation.

Referring to this kind of example:

```
function rrule(::typeof(foo), a, b)
    y = foo(a, b)
    h = buzz(a, b)
    ∂a = Rule(ȳ -> 2h)
    ∂b = Rule(ȳ -> 3h)
    return ∂a, ∂b
    return y, (∂a, ∂b)
end
```

The original justification was that if you didn't want `buzz` to be called at all (i.e. you don't want derivatives at all), why call `rrule` at all (as opposed to `foo(a, b)`)? This justification is somewhat broken in cases where the caller doesn't have enough information to make that determination, though. More importantly, I just realized that this justification is actually more broken (or would at least lead to more hoop-jumping) for ternary (and higher arity) functions where you'd like to stage different computations between e.g. different pairs of Rules.

> I am not sure I really understand what you mean when you say
that frule and rrule are duels, to help support mixed-mode AD.
I don't really know that stuff, to be completely honest.
I don't think we would have broken any elegence because we turned both
rrules and frules use of rules inside out.
So they should still match-up.

For this kind of thing, unary functions aren't good test cases; try coming up with a toy function (scalar is fine) of e.g. input arity 3 and output arity 2. That's usually enough to get a sense of how input/output w*ggles have to be munged to transpose a rule. I just played around with it a bit and I *think* things seem fine from the mixed-mode perspective. AFAICT `frule`/`rrule` still relate to in a correct way, it's just that the "locus" of the consistency shifted a bit.

I'm fully in favor of this now, thanks again for the examples 👍

oxinabox commented:
> The original justification was that if you didn't want buzz to be called at all (i.e. you don't want derivatives at all), why call rrule at all (as opposed to foo(a, b))? This justification is somewhat broken in cases where the caller doesn't have enough information to make that determination, though..

Yes, I think that case is common in the general autodiff setting.
Where it isn't unreasonably to basically `overdub(ctx, f:: Function, xs...) =frule(f, xs...)` (not literally that but something that 
Does that call internally)

Incontrast to what I have been thinking of as the restricted structure setting (e.g. Mocha.jl, JuMP, and ModelingToolKit.jl) where we can predetermine exactly what variables are being differncated w.r.t what others, because the DSL is restricted


>  More importantly, I just realized that this justification is actually more broken (or would at least lead to more hoop-jumping) for ternary (and higher arity) functions where you'd like to stage different computations between e.g. different pairs of Rules.

I think it still is. 


> For this kind of thing, unary functions aren't good test cases; try coming up with a toy function (scalar is fine) of e.g. input arity 3 and output arity 2. 

Yeah, that is a good idea.
There are none in ChainRules right now.


I think one real world case that will come up is defining `rrule(::typeof(rrule), ::typeof(foo), xs, ys)`
Which would really put stuff through its paces, and which I've been asked about.

tkf commented:
> I'm fully in favor of this now

@jrevels It seems you were questioning the justifications in the middle paragraph. Could you explain why did you still think the current PR is a good direction?  Sorry if I misread your post.

@oxinabox The body of pullback function seems to be a bad location to do any non-trivial computation.  For example, you wouldn't want to do

(following the advice of @jrevels about input arity 3)

```julia
function rrule(::typeof(foo), a, b, c)
    y = foo(a, b, c)
    function foo_pullback(ȳ)
        h = buzz(ȳ, a, b)

        ∂a = @thunk(2h)
        ∂b = @thunk(2h)
        ∂c = @thunk(bar(ȳ))
        return NO_FIELDS, ∂a, ∂b, ∂c
    end
    return y, foo_pullback
end
```

as calling `buzz` wastes space and time if you only need `∂c`.  I think, even after this PR, `buzz` should be wrapped in some kind of caching thunk https://github.com/JuliaDiff/ChainRulesCore.jl/issues/7 to avoid computation as much as possible.

```julia
function rrule(::typeof(foo), a, b, c)
    y = foo(a, b, c)
    function foo_pullback(ȳ)
        h = @cached_thunk buzz(ȳ, a, b)

        ∂a = @thunk(2extern(h))
        ∂b = @thunk(2extern(h))
        ∂c = @thunk(bar(ȳ))
        return NO_FIELDS, ∂a, ∂b, ∂c
    end
    return y, foo_pullback
end
```

I know that you would like to avoid boxing and the above approach seems to be against it.  But I think it can be avoided if you allocate an empty buffer in the pullback function and then reinterpret in each thunk; i.e., something like

```julia
buf = Vector{UInt8}[]
h = @thunk begin
    if isempty(buf)
        do computation with buf
    end
    return reinterpreted buf
end
```

but with a bit more automation.

@oxinabox @jrevels I thought the motivation of this PR was more related to #8.  That is to say, if you want to differentiate (say) `x -> f((a=x, b=y), z)` where `y` and `z` are constants, it would be nice to be able to evaluate pullback for `x`, `y`, and `z` separately (possible if they are all thunk) rather than just `(a=x, b=y)` and `z` separately (pre-PR/current status of ChainRulesCore.jl).  In other words, lazy evaluation should always happen at "leafs."

oxinabox commented:
> @oxinabox The body of pullback function seems to be a bad location to do any non-trivial computation. For example, you wouldn't want to do

Yes, as I commented, this does not solve the 3+ airity, shared by two case.
It solves the 2 airity, not shared with main result, case.
Fortunately the case is not too common (except for the NO_FIELDS return which is ubiqitious 😟, still in usage i hope that doesn't end up mattering since it is the result that noone uses normally so its evaluation of extra stuff to get it isn't a problem (if it were the other way round that would be a problem, where you do lots of work for the side that noone uses))
Ideally, the compiler would optimize out the nonused pure function calls.
(Which AFAICT it only does for fully constant folded functions right now?)

>  I think, even after this PR, buzz should be wrapped in some kind of caching thunk #7 to avoid computation as much as possible.

Yes, this remains known.
It is nontrivial, the idea of reinterpretting a buffer is interesting.
But lets keep that discussion in #7 

> @oxinabox @jrevels I thought the motivation of this PR was more related to #8. That is to say, if you want to differentiate (say) x -> f((a=x, b=y), z) where y and z are constants, it would be nice to be able to evaluate pullback for x, y, and z separately (possible if they are all thunk) rather than just (a=x, b=y) and z separately (pre-PR/current status of ChainRulesCore.jl). In other words, lazy evaluation should always happen at "leafs."

This PR  more paves the way for that being cleaner and more consistent than solves it outright.
Since it means that leafs at the top and leafs at the bottom are the same (since no more Rule layer).
But this PR itself doesn't fully change enough about how differentials work to support AbstractDifferentials in deeper leaves.
Definately we want that though, but it will be another PR. This one is already *huge* (when you look also at its partner)
and changes a bunch of things.

oxinabox commented on test/rules.jl:
```diff
@@ -29,10 +29,123 @@ _second(t) = Base.tuple_type_head(Base.tuple_type_tail(t))
+end
+
+
+@testset "Basic Wirtinger scalar_rule" begin
```
These came out of what uses ti be in `tests/rule_types.jl`

oxinabox commented:
Functionality all done on both this and https://github.com/JuliaDiff/ChainRules.jl/pull/91
everything is passing tests when used together.

TODO:
 - resolve all open discussions from current reviews
 - write enough introductory documentation (mainly a terminology section) so more people can review
 - write a some stuff expressly about what to look for when reviewing
 - fix the things they find when they review.

nickrobinson251 commented:
Thanks for all the work on this! I will be able to review when I am back on Monday :)

tkf commented:
@oxinabox Congratulations and thanks a lot for this RP!

> Yes, as I commented, this does not solve the 3+ airity, shared by two case.

Sorry, I think I misread the comment.

> Fortunately the case is not too common

I think the problem can appear even with arity < 3 when you have sub-computation with variables not involving the differentiation.  For example, if you want to take the derivative of `x -> x * f(y, z)`, you do not want to evaluate the pullbacks for `y` and `z`, right?  However, if adjoint of `f` eagerly compute the common computation for `y` and `z` (like `buzz` in your example), it would be a waste of time.  Of course, an AD engine can be sufficiently clever to not evaluate anything related to `f(y, z)` in the backward pass, but I've got an impression that this is not commonly implemented in Julian ADs from the discussion with @willtebbutt in https://github.com/FluxML/Zygote.jl/issues/323.  But I'm not entirely sure how this part of Zygote or any other Julian AD works.  I would be happy to know if my concern is invalid.

> Ideally, the compiler would optimize out the nonused pure function calls.

I think machine learning is one of the biggest use-case of ChainRules, right?  In this case, many of the computation would allocate arrays and I suppose the compiler is not clever enough (yet?) to elide them.  It'd be nice if ChainRules(Core) does not rely on dead code elimination too much in design.

oxinabox commented:
> I think the problem can appear even with arity < 3 when you have sub-computation with variables not involving the differentiation. For example, if you want to take the derivative of x -> x * f(y, z), you do not want to evaluate the pullbacks for y and z, right? However, if adjoint of f eagerly compute the common computation for y and z (like buzz in your example), it would be a waste of time. 
> Of course, an AD engine can be sufficiently clever to not evaluate anything related to f(y, z) in the backward pass, but I've got an impression that this is not commonly implemented in Julian ADs from the discussion with @willtebbutt in FluxML/Zygote.jl#323. But I'm not entirely sure how this part of Zygote or any other Julian AD works. I would be happy to know if my concern is invalid.

This seems like a problem for the AD tool, not for ChainRules.
Note that without this PR the common computation for `y` and `z` is done in the body of the `rrule`.
Which means to avoid it not only does the AD tool have to avoid triggering the pullback,
it needs to avoid calling `rrule` in the first place.
Which is at least (maybe exactly as hard) as knowing not to call the pullback.

The correct answer is to not call the pullback,
you know in advance that for something that does not occur in the arguments (including the implict `f` itself),
that there will be no corresponding partial coming back.
There is no index to look in to get the dervivative of interest.
So I don't see how the pullback would be used once invoked.
idk how Zygote works in this respect.


> > Ideally, the compiler would optimize out the nonused pure function calls.
> 
> ...In this case, many of the computation would allocate arrays and I suppose the compiler is not clever enough (yet?) to elide them. It'd be nice if ChainRules(Core) does not rely on dead code elimination too much in design.

Yes, this is all true and verified.  
I still think not returning multiple rules is worth it.
The only common case to have shared work for 2 of 3 partials right now is the
self-arg case, which is NO_FIELDS for any nonclosure/nonfunctor.
Which if one knows one is going to query only for that and it turns out to be a bottleneck,
then one can check beforehand if the `f` is a closure/functor or not and avoid it.
Once we solve the n-of-m shared work problem we will solve this too.

simeonschaub commented on src/differentials.jl:
```diff
@@ -227,3 +245,62 @@ end
 Base.show(io::IO, x::Thunk) = println(io, "Thunk($(repr(x.f)))")
+
+"""
+    InplaceableThunk(val::Thunk, add!::Function)
```
Sorry for being so pedantic about names, but `InplaceableThunk` just reads a little bit clunky for me :upside_down_face: . What about just `InplaceThunk`?

oxinabox replied:
Problem is it isn't always inplace.
Only of you do `accumulate!` is it inplace.

simeonschaub replied:
Even if it isn't always inplace, I would say `InplaceThunk` still would signify that it can be used inplace, but doesn't necessarily mean that it has to. Is `Thunk!` punning too hard?

simeonschaub commented:
This is really looking awesome! Great effort!

jrevels commented:
> @jrevels It seems you were questioning the justifications in the middle paragraph. Could you explain why did you still think the current PR is a good direction? Sorry if I misread your post.

(@tkf just to clarify: When I said "The original justification..." in my previous post, I was referring to my own memories of how I justified the original `Rule` design, not Lyndon's justifications. In other words, I was pointing out where my justifications for the original design were iffy :) Sorry for the confusion!)

tkf commented:
@oxinabox Many of my points are in favor with this PR.  I just wanted to pointed out that lazy shared evaluation alone may not be as strong as you pitched, just in case you want to evaluate the direction of this PR carefully.  I'll continue other discussions elsewhere (esp. caching).

@jrevels Thanks, that clarifies a lot.  I was indeed totally misreading your post.

nickrobinson251 commented:
Having seen both options, I'm convinced `InplaceableThunk` is the better/clearer name

oxinabox commented:
Ok after talking to @nickrobinson251  and looking at how it looks.
`InplaceableThunk` seems better and more clear,
it can't be misinterpretted as a verb.

Also gone with `refine_differential` to capture refininga Wirtinger on the real-domain.

I think this is done.

oxinabox commented:
Rebased, and squashed some of them.
Going to try shuffling the commits to squash it down some more.

Normally I am hesitant to squash during PR review but this has had a lot of review so far,
so making each commit distinct in purpose seem apppropriate now

oxinabox commented:
All tests (except inegration tests) are passing.

Shuffle rebasing is hard, not sure if worth it

nickrobinson251 commented on src/differentials.jl:
```diff
@@ -199,31 +218,87 @@ Thunk(var"##8#10"())
+- The expression wrapping something in a `struct`, such as `Adjoint(x)` or `Diagonal(x)`
+- The expression being a constant
+- The expression being itself a `thunk`
+- The expression being from another `rrule` or `frule` (it would be `@thunk`ed if required by the defining rule already)
```
This entire section is great

Shall we move it to a page in the docs?

nickrobinson251 commented on src/differentials.jl:
```diff
@@ -199,31 +218,87 @@ Thunk(var"##8#10"())
+(Otherwise one can just use a normal `Thunk`).
+
+Most operations on an `InplaceableThunk` treat it just like a normal `Thunk`;
+and destroy its inplacability.
```
```suggestion
and destroy its ability to work inplace.
```

nickrobinson251 commented on src/differentials.jl:
```diff
@@ -199,31 +218,87 @@ Thunk(var"##8#10"())
-Base.conj(x::Thunk) = @thunk(conj(extern(x)))
+# The real reason we have this:
+accumulate!(Δ, ∂::InplaceableThunk) = ∂.add!(Δ)
+store!(Δ, ∂::InplaceableThunk) = ∂.add!((Δ.*=false))  # zero it, then add to it.
```
```suggestion
store!(Δ, ∂::InplaceableThunk) = ∂.add!((Δ .*= false))  # zero it, then add to it.
```

nickrobinson251 commented on src/operations.jl:
```diff
@@ -0,0 +1,45 @@
+Similar to [`accumulate`](@ref), but attempts to compute `Δ + rule(args...)` in-place,
+storing the result in `Δ`.
+
+Note: this function may not actually store the result in `Δ` if `Δ` is immutable,
```
```suggestion
!!! note
    this function may not actually store the result in `Δ` if `Δ` is immutable,
```

nickrobinson251 commented on src/operations.jl:
```diff
@@ -0,0 +1,45 @@
+"""
+    store!(Δ, ∂)
+
+Stores `∂`, in `Δ`, overwriting what ever was in `Δ` before.
```
```suggestion
Stores `∂` in `Δ` overwriting whatever was in `Δ` before.
```

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,90 @@
+
+This is able to deal with fairly complex expressions for `f`:
+
+    julia> propagator_name(:bar, :pushforward)
```
any reason not to use `jldoctest` here?

oxinabox replied:
Not exported

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,90 @@
+Returns the expression for the propagation of
+the input gradient `Δs` though the partials `∂s`.
+
+𝒟 is an expression that when evaluated returns the type-of the input domain.
```
```suggestion
𝒟 is an expression that when evaluated returns the type of the input domain.
```

nickrobinson251 commented on src/rule_definition_tools.jl:
```diff
@@ -1,5 +1,90 @@
+function standard_propagation_expr(Δs, ∂s)
+    # This is basically Δs ⋅ ∂s
+
+    # Notice: the thunking of `∂s[i] (potentially) saves us some computation
```
```suggestion
    # Notice: the thunking of `∂s[i]` (potentially) saves us some computation
```

nickrobinson251 commented on src/rules.jl:
```diff
@@ -4,7 +4,7 @@
 #=
 In some weird ideal sense, the fallback for e.g. `frule` should actually be "get
-the derivative via forward-mode AD". This is necessary to enable mixed-mode
+the derivative via forward-ode AD". This is necessary to enable mixed-mode
```
```suggestion
the derivative via forward-mode AD". This is necessary to enable mixed-mode
```
